subprojects/core/src/main/java/org/gradle/api/internal/changedetection/changes/RebuildIncrementalTaskInputs.java

```diff
@@ -29,20 +32,37 @@
 public class RebuildIncrementalTaskInputs extends StatefulIncrementalTaskInputs {
     private static final Logger LOGGER = LoggerFactory.getLogger(RebuildIncrementalTaskInputs.class);
 
-    private final FileCollection inputFiles;
+    private final CurrentTaskExecution execution;
 
-    public RebuildIncrementalTaskInputs(Task task, TaskProperties taskProperties) {
+    public RebuildIncrementalTaskInputs(Task task, CurrentTaskExecution execution) {
         LOGGER.info("All input files are considered out-of-date for incremental {}.", task);
-        this.inputFiles = taskProperties.getInputFiles();
+        this.execution = execution;
     }
 
     public boolean isIncremental() {
         return false;
     }
 
-    public void doOutOfDate(Action<? super InputFileDetails> outOfDateAction) {
-        for (File file : inputFiles) {
-            outOfDateAction.execute(new RebuildInputFile(file));
+    public void doOutOfDate(final Action<? super InputFileDetails> outOfDateAction) {
+        PhysicalSnapshotVisitor visitor = new PhysicalSnapshotVisitor() {
+            @Override
+            public boolean preVisitDirectory(PhysicalDirectorySnapshot snapshot) {
+                return true;
+            }
+
+            @Override
+            public void visit(PhysicalSnapshot snapshot) {
+                outOfDateAction.execute(new RebuildInputFile(new File(snapshot.getAbsolutePath())));
+            }
+
+            @Override
+            public void postVisitDirectory(PhysicalDirectorySnapshot snapshot) {
+                // No-op
+            }
+        };
+
+        for (CurrentFileCollectionFingerprint fingerprint : execution.getInputFingerprints().values()) {
+            fingerprint.visitRoots(visitor);
```

Comment:
Please use `fingerprint.getFingerprints()` for the stuff you want to pass to the `outOfDate` action, not the `visitRoots` method. This makes sure that the correct output filtering already happened. Then you also won't need to use a visitor.